ddht/v5/crawl.py

```diff
@@ -0,0 +1,234 @@
+import contextlib
+import logging
+from typing import Iterator, List, Set
+
+from eth_enr import ENRAPI, ENRDB, ENRManager, default_identity_scheme_registry
+from eth_enr.exceptions import OldSequenceNumber
+from eth_typing import NodeID
+from eth_utils import encode_hex
+import trio
+
+from ddht.app import BaseApplication
+from ddht.boot_info import BootInfo
+from ddht.constants import DEFAULT_LISTEN, IP_V4_ADDRESS_ENR_KEY
+from ddht.exceptions import UnexpectedMessage
+from ddht.v5.app import get_local_private_key
+from ddht.v5.client import Client
+from ddht.v5.messages import FindNodeMessage
+
+logger = logging.getLogger("ddht.crawler")
+
+
+class StopScanning(Exception):
+    pass
+
+
+class Crawler(BaseApplication):
+    def __init__(self, concurrency: int, boot_info: BootInfo) -> None:
+        super().__init__(boot_info)
+        self.concurrency = concurrency
+
+        self.enr_db = ENRDB(dict(), default_identity_scheme_registry)
+
+        self.sock = trio.socket.socket(
+            family=trio.socket.AF_INET, type=trio.socket.SOCK_DGRAM
+        )
+
+        self.private_key = get_local_private_key(boot_info)
+
+        self.enr_manager = ENRManager(private_key=self.private_key, enr_db=self.enr_db)
+        self.enr_manager.update((b"udp", boot_info.port))
+
+        my_node_id = self.enr_manager.enr.node_id
+        self.client = Client(self.private_key, self.enr_db, my_node_id, self.sock)
+
+        self.active_tasks = ActiveTaskCounter()
+
+        self.seen_nodeids: Set[NodeID] = set()
+        self.responsive_peers: Set[NodeID] = set()
+
+        self.enrqueue_send, self.enrqueue_recv = trio.open_memory_channel[ENRAPI](
+            10_000
+        )
+
+    async def fetch_enr_bucket(self, remote_enr: ENRAPI, bucket: int) -> List[ENRAPI]:
+        peer_id = remote_enr.node_id
+
+        logger.debug(f"sending FindNode. nodeid={encode_hex(peer_id)} bucket={bucket}")
+
+        try:
+            with trio.fail_after(2):
+                find_node = FindNodeMessage(
+                    request_id=self.client.message_dispatcher.get_free_request_id(
+                        peer_id
+                    ),
+                    distance=bucket,
+                )
+
+                responses = await self.client.message_dispatcher.request_nodes(
+                    peer_id, find_node
+                )
+
+                return [enr for response in responses for enr in response.message.enrs]
```

Comment:
For consistency with our other code this should probably be converted to a tuple comprehension `tuple(thing for thing in things)`